src/dsp_tools/commands/xmlupload/check_consistency_with_ontology.py

```diff
@@ -0,0 +1,74 @@
+# sourcery skip: use-fstring-for-concatenation
+
+from lxml import etree
+
+from dsp_tools.commands.xmlupload.ontology_client import OntologyClient, format_ontology
+from dsp_tools.commands.xmlupload.ontology_diagnose_models import OntoCheckTool, Ontology, UnknownOntologyElements
+from dsp_tools.models.exceptions import BaseError
+
+
+def do_xml_consistency_check(onto_client: OntologyClient, root: etree._Element) -> None:
+    """
+    This function takes an OntologyClient and the root of an XML.
+    It retrieves the ontologies from the server.
+    It iterates over the root.
+    If it finds any unknown properties or classes, they are printed out and a UserError is raised.
+
+     Args:
+         onto_client: client for the ontology retrieval
+         root: root of the XML
+
+     Raises:
+         UserError: if there are any unknown properties or classes
+    """
+    onto_tool = OntoCheckTool(
+        default_ontology_prefix=onto_client.default_ontology,
+        onto_lookup=_get_project_and_knora_ontology_from_server(onto_client),
+        save_location=onto_client.save_location,
+    )
+    _iterate_over_xml_find_problems(root, onto_tool)
+
+
+def _get_project_and_knora_ontology_from_server(onto_client: OntologyClient) -> dict[str, Ontology]:
+    ontologies = onto_client.get_all_ontologies_from_server()
+    return {onto_name: format_ontology(onto_graph) for onto_name, onto_graph in ontologies.items()}
+
+
+def _iterate_over_xml_find_problems(root: etree._Element, onto_tool: OntoCheckTool) -> None:
+    unknown_eles = UnknownOntologyElements(save_path=onto_tool.save_location)
+    for resource in root.iterchildren():
+        if not _diagnose_classes(resource.tag, onto_tool):
+            unknown_eles.classes.append((resource.attrib["id"], resource.attrib["restype"]))
+        if problems := _diagnose_all_properties(resource, onto_tool):
+            unknown_eles.properties.extend(problems)
+    if unknown_eles.not_empty():
+        unknown_eles.execute_problem_protocol()
```

Comment:
Regarding the first part.
I did a version of this in the beginning and it became overly complicated. But if I were to extract it in the tuple format then maybe I could cleanly separate this. Thanks!

Reply:
> Also, I'm not a fan of creating an object and then modifying its members. How about creating lists first and then creating the object with the finished lists? Then it can also be a frozen object.

Is there a specific reason?

Reply:
Generally preferring immutability is also my main reason. But there is also a cosmetic aspect to it, namely the accessing of the variable is not nested. just to illustrate (notice the comments):

```python
@dataclass
class Something:
    a: list[int]
    b: list[int]

def mutable() -> Something:
    something = Something()
    for i in range(99):
        something.a.append(i+1)  # compare this line
    for i in range(99):
        something.b.append(i * i)
    return something

def immutable() -> Something:
    a = []
    for i in range(99):
        a.append(i+1)  # with this line
    b = []
    for i in range(99):
        b.append(i * i)
    return Something(a, b)
```

of course this is a stupid example because this would normally be solved with list comprehension... but you get my point